### ISSUE TITLE
Files in folder are not visible to clients

### ISSUE THREAD

**Issue by xzenor** [projectsend/projectsend#1333]
I, as a user added a bunch of files and gave them the proper permissions. Files were visible by the clients.
Then created a folder and moved the files in there. Files are now invisible by the clients. I checked the permissions and the files still have the correct permissions. the clients don't see folders at all.  

Am I missing something? Did I mess something up with upgrading?  

NginX-1.26 and php8.2
I'd be happy to provide more info if I know what you need :)

**Comment by ignacionelson** [comment]
Sorry for the issue!
Clients do not see any folder? Or is the entire breadcrumbs/folder block no there?

**Comment by xzenor** [comment]
Breadcrumbs are there.. there's just no folder to open..
![image](https://github.com/user-attachments/assets/1b3d3f24-a4f4-45cb-a877-9a72ae373e7f)

**Comment by xzenor** [comment]
Upgraded to php-8.3 but no difference. I use firefox normally and tried it in Edge but same issue.  

Here's the console log from Edge's Developer options.  

![image](https://github.com/user-attachments/assets/7d87fe78-e0ba-4d75-8588-8ccb813c553a)

**Comment by ignacionelson** [comment]
please try this

https://docs.projectsend.org/about/troubleshooting/debug-mode

It's strange that at least the root folder is not appearing

**Comment by xzenor** [comment]
Just some deprecation warnings.. No errors  

![image](https://github.com/user-attachments/assets/bc6c9950-a15c-4b38-aa22-044a007264a6)

**Comment by xzenor** [comment]
I'm a bit confused about your previous comment.. what did you mean with the root folder not appearing? Anything not in folders is visible, and that is the root, isn't it? As long as I don't put files in folders then it just works. Just like in the previous version.

**Comment by ignacionelson** [comment]
Sorry, I'm still trying to reproduce this

**Comment by ignacionelson** [comment]
Do you by chance have any php 7.x installed to test this?

**Comment by mikkeljohnsen** [comment]
I use the container image, and I see the same issue. The clients can't se the folder.

And there is no way to manage folders. Not able to delete them again and assign access.

Version r1720

**Comment by ignacionelson** [comment]
@mikkeljohnsen you should be able to delete them by right-clicking them. I know the UI is not ideal, it's just a base to build from in the future.

**Comment by mikkeljohnsen** [comment]
@ignacionelson Thanks it works with right clicking.

So now I just need to be able to assign the folders to a client/client group.

**Comment by xzenor** [comment]
> Do you by chance have any php 7.x installed to test this?

No, sorry. It's not available for install anymore or I would've tried it.   

I did try using a clean install with a clean database (did copy the nginx config from the live one) and same issue.. So I guess we can rule out the upgrade path and the database. So I would guess it had to be something on my system. a permission, a missing extension or an option in one.. But @mikkeljohnsen used the container image and has the same issue..   

I'm a bit lost in pinpointing a cause. What information do you need that would help?

**Comment by mikkeljohnsen** [comment]
It is also very difficult to see what subfolders belongs to. In the Location it shows parent folders and subfolder on same level (not indented).

**Comment by mikkeljohnsen** [comment]
The problem with visible folders is, that a folder is normally created by the manager, and will then be owned by that maneger. the table "tbl_folders" will have "user_id=1".

If I change the "user_id" to the actual user, then it shows up.

And the administrator can still see the folder.

**Comment by xzenor** [comment]
> The problem with visible folders is, that a folder is normally created by the manager, and will then be owned by that manager. the table "tbl_folders" will have "user_id=1".
> 
> If I change the "user_id" to the actual user (client), then it is visible for the client. And the administrator can still see the folder.
> 
> We need a "tbl_folders_relations"

aaaah yes! Once I change the owner of the folder in the database then that specific client can see it! But there's no way to change these permissions in the web-interface, is there?

I'm really confused how this is supposedly working for other people. There is a difference between users and clients though. Even a user with 'only' the upload role can see all folders so users are fine, but only clients that have their id set as the owner of the folder (via a db change) can see the folder.

**Comment by ignacionelson** [comment]
I see. Thanks for taking the time to write your message. But folder visibility is not dictated by its owner. The condition is that it must contain a file visible to that client. I wonder if the problem is that the client has a file in a subfolder and not in the parent (I need to check the code to make sure this doesn't happen, but I'm not able to do that ATM)

**Comment by ignacionelson** [comment]
Please keep in mind that this is a first release, I'll be adding more functionality to folders as soon as I find the time. I need to make an UI to edit permissions (public-private, sharing entire folder, etc)

**Comment by mixylplik3** [comment]
I am having this issue. Breadcrumbs are there, but any files I move to a folder just disappear. The dividers under the breadcrumb and above "Found # elements" are there, but it's empty space between them where the folders should go and the HTML has some code that isn't populating. PHP 8.2.22

VERY excited about this feature. Huge upgrade for my use case. 

![Screenshot 2024-08-20 at 2 34 53 PM](https://github.com/user-attachments/assets/e0e92e4b-e3c6-436f-8d56-10272eb519f1)

**Comment by xzenor** [comment]
> I see. Thanks for taking the time to write your message. But folder visibility is not dictated by its owner. The condition is that it must contain a file visible to that client. I wonder if the problem is that the client has a file in a subfolder and not in the parent (I need to check the code to make sure this doesn't happen, but I'm not able to do that ATM)

Oh the file is definitely visible by the client. Until I move it to a folder. If I move it back to the root it's visible again. move it into the folder and it's gone (and also no folder visible). If I do change the owner directly in SQL then it works and the folder is visible with the files in it.  

> Please keep in mind that this is a first release, I'll be adding more functionality to folders as soon as I find the time. I need to make an UI to edit permissions (public-private, sharing entire folder, etc)  

Can't speak for everyone but I want to make very clear that I just want to help fix this bug. There's no anger. Not mad at someone. not annoyed. No hard feelings. and my sincerest apologies if I ever made you feel that there was. I'm just excited about the feature. It's really a great step forwards and I would love to pinpoint the cause of the problem so you can build an even better product (and I can then use that :P). I'm fully aware it's a first release and it probably has some teething issues to work out. Which is fine.

**Comment by ignacionelson** [comment]
> and my sincerest apologies if I ever made you feel that there was.

Absolutely not! I typed that message while in the middle of other work-related issues so it came out wrong. Sorry!
The thing is that I've delayed the release for a year, it was VERY hard to find real time to work on PS for a while. So when I came back to it, I guess my testing was not enough. I should have delayed the release a little longer probably!

**Comment by ignacionelson** [comment]
Indeed, on a new install I'm having these issues as well. I guess I need to rethink a big part of this.
Reworking the logic so folders do not depend on contents, but on its own permissions...
Having folders visibility depending on files meant that you could rearrange files without having to worry about folders permissions.

**Comment by Todilo** [comment]
Any workaround to this issues, folders is essential to our needs and I did a fresh setup (docker) which does not reveal folders or breadcrumbs for those despite having files allocated to a group and its users.

**Comment by Matani-Git** [comment]
Fix for Folder Visibility and Client Access Control

Modified the Folders class to properly handle folder visibility for clients while maintaining existing functionality for admins/operators. The main changes address an issue where clients could see folders they shouldn't have access to.

Key Changes:
1. Added user level detection from tbl_users to automatically determine client/admin status
2. Revised folder visibility logic for clients (level = 0) to:
   - Only show folders containing files explicitly assigned to them
   - Show public folders
   - Remove the user_id filter for clients which was causing incorrect folder visibility

3. Modified the SQL query structure to:
   - Use EXISTS clause to check file relations
   - Properly handle folder hierarchy
   - Maintain performance with optimized joins
   - Keep all existing functionality (search, public folders, etc.)

The fix ensures that:
- Clients only see folders containing their assigned files
- Public folders remain visible to all users
- Admins/operators maintain their existing folder access
- Folder hierarchy and navigation remain intact
- Existing features (search, filtering, etc.) continue to work as before

Impact:
This change improves security by properly restricting folder visibility while maintaining a seamless user experience. No database schema changes are required, and the fix is backwards compatible with existing functionality.

Usage remains the same - just pass the user_id to getFolders():
$folders = $foldersObj->getFolders([
    'user_id' => $current_user_id,
    'parent' => $parent_folder_id  // optional
]);

the folders.php class is attached to test!
[Folders.php.zip](https://github.com/user-attachments/files/17624303/Folders.php.zip)

**Comment by k0d3g3ar** [comment]
Yes, this is definitely an issue.  The fix is to change the owner of the folders in the tbl_folders table.  I created my system so that I would be the master administrator.  I upload the files and originate the content.  Then I have user accounts that restrict what they can see.  By using folders, I thought it would organize things for all users.  However the folders appear to be owned and only visible by the creator.  So I can see things organized that way, but my users cannot.

Once I changed the owner ID to the user ID, they could see them but as the administrator, I could still see all the folders anyway.  So despite the manual override, this works out great in use.  But a pain to administer.

**Comment by Matani-Git** [comment]
I have fixed the Folders Class Issues for me, i hope this will be fixed soon in the develop rep. and next release

[ignacionelson](https://github.com/ignacionelson) thank you for that great work :)

To test my folders.php fix just backup the folders.php under include/classes/ and replace it with the attached folders.php

Fix description
Several  bugs in the Folders class have been fixed, improving folder visibility and hierarchy management for clients:

- Client Folder Display: Fixed issue where clients couldn't see parent folders when files were assigned in subfolders. Implemented recursive folder traversal to ensure proper visibility of the entire folder path when a client has access to files in any subfolder.

- Folders are now automatically visible to clients only if they contain files assigned to them or are parent folders of such files.

-Missing Folder Relations: Fixed null folder_id values in tbl_files_relations table with an automatic update query:

UPDATE `tbl_files_relations` r 
SET folder_id = (SELECT folder_id FROM tbl_files fi WHERE fi.id = r.file_id)

This ensures proper folder relations for all file assignments.

-Empty Parent Folders: Fixed issue where parent folders weren't visible if they didn't contain directly assigned files. The folder hierarchy is now properly maintained and displayed, showing all relevant parent folders regardless of direct file assignments.

The updated class uses a recursive query approach to properly handle folder hierarchies while maintaining security. Clients now see the complete folder structure for their assigned files without exposing unrelated folders.
[Folders.php.zip](https://github.com/user-attachments/files/17685567/Folders.php.zip)

**Comment by Matani-Git** [comment]
I have fixed the Folders Class Issues for me, i hope this will be fixed soon in the develop rep. and next release

To test my folders.php fix just backup the folders.php under include/classes/ and replace it with the attached folders.php


here is the fixed folders.php :
[Folders.php.zip](https://github.com/user-attachments/files/17685634/Folders.php.zip)

Fix description
Several  bugs in the Folders class have been fixed, improving folder visibility and hierarchy management for clients:

- Client Folder Display: Fixed issue where clients couldn't see folders and parent folders when files were assigned in folders and subfolders. Implemented recursive folder traversal to ensure proper visibility of the entire folder path when a client has access to files in any subfolder.

- Folders are now automatically visible to clients only if they contain files assigned to them or are parent folders of such files.

- Missing Folder Relations: Fixed null folder_id values in tbl_files_relations table with an automatic update query:

UPDATE `tbl_files_relations` r 
SET folder_id = (SELECT folder_id FROM tbl_files fi WHERE fi.id = r.file_id)

This ensures proper folder relations for all file assignments.

- Empty Parent Folders: Fixed issue where parent folders weren't visible if they didn't contain directly assigned files. The folder hierarchy is now properly maintained and displayed, showing all relevant parent folders regardless of direct file assignments.

The updated class uses a recursive query approach to properly handle folder hierarchies while maintaining security. Clients now see the complete folder structure for their assigned files without exposing unrelated folders.

**Comment by Matani-Git** [comment]
The Projectsend release r1720 introduced the folder implementation. However, this feature is not very useful if clients cannot see the folders and their contents!

Until now, we have only used categories and category filters to reduce the file lists. With the new folder feature, the structure has become even better!

The folders in Projectsend are not traditional directories but virtual folders designed to improve organization for admins and clients. All files remain in the one directory, and the folder information is stored in the tbl_folders table.

**Problem**: When I upload new files as an admin or operator and assign these files to clients and client groups, the clients cannot see the files in the designated folders. They can only see files in the root folder.

Analysis: In the getFolders function within the Folders class, the publicorclient argument is not set. As a result, folders that contain files assigned to specific clients or client groups are not displayed.

**Solution:**
Check the access level (level): if level = 0, it indicates a client; otherwise, it’s an admin or operator.
If level ≠ 0, keep the behavior as it is.
If level = 0, retrieve the client ID. Display all folders that contain files assigned to the client or to a group of which the client is a member.
Set the folders as “public” so that clients can see the contents. The permissions remain attached to the files themselves, not to the folders. (I am not sure how the future folder functionality is planned, but this setup is perfect for our needs. I hope this will be available as an option in a future release.)
Empty folders or folders without assigned files will be hidden from the client. If a client has a file in a subfolder, all parent directories up to the root will be rendered and shown.

**Fix**:
The fix involves modifying the Folders class in projectsend/includes/Classes/Folders.php. Rename the old Folders.php file to Folders.php.bak, upload the new file, and test.

You can download the fix here: [Folders.zip](https://github.com/user-attachments/files/17704311/Folders.zip)

**Comment by ignacionelson** [comment]
@Matani-Git thank you! This is great!!
Would you like to submit a pull request so you can be listed as contributor on the repo?

**Comment by stefan-thaler-compLex** [comment]
@Matani-Git Thanks for the fix, but I think it's not working with MySQL Version 5.7, because I'm getting following error:
"Fatal error: Uncaught PDOException: SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'RECURSIVE folder_hierarchy AS ( -- Base case: Get folder' at line 24 in..."

**Comment by Matani-Git** [comment]
Hi @stefan-thaler-compLex The error is occurring because MySQL 5.7 doesn't support Common Table Expressions (CTEs) with the WITH RECURSIVE syntax. CTEs were introduced in MySQL 8.0. However, i have rewrite this query to work with MySQL 5.7 using a different approach.

here is the fixed folders.php  [Folders.php.zip](https://github.com/user-attachments/files/17757539/Folders.php.zip)

**Comment by xzenor** [comment]
> @Matani-Git Thanks for the fix, but I think it's not working with MySQL Version 5.7, 

MySQL 5.7 was EOL a little over a year ago. It's time to upgrade. It has known security vulnerabilities.

**Comment by stefan-thaler-compLex** [comment]
Hi @Matani-Git,
you are my hero, thanks for the quick fix - I tested it and it works now.

**Comment by mbiolsi** [comment]
The patch DID fix the disappearing folders! Thank you. 

However, there is another consideration. IF a client has the ability to upload files, they can see ALL folders, even if they do not have any files in some of the folders. 

REPLICATE: 
* log in as client
* upload new file
* on the edit files page select a location from the dropdown - you will see all folders. 

![image](https://github.com/user-attachments/assets/8f62e1fa-accb-4ff6-8f62-404f95e005cc)

Is there a way to make sure clients only see folders assigned to them?

**Comment by ignacionelson** [comment]
@Matani-Git did a far better job than I could have done with this fix. Maybe you can provide your input on this issue? Thank you!

**Comment by mbiolsi** [comment]
FIRST, I love the addition of folders. Thanks! I am new to projectsend and cannot imagine it without folders. 

Some thoughts on permissions and folders:

Permissions are never simple. The method you have chosen of assigning permissions to a file makes a lot of sense. It is very easy to manage when adding new files.  I LIKE that the client can see the parent folder as long as they have a file in a subfolder. This is super helpful for navigation. However, that means that a series of checks need to be in place to SHOW or HIDE folders based upon the files in them. I have not looked at the code, but it seems like that could be a lot of file checking as the file list gets longer. But the code would have to be able see that a client has access to a file in a folder and give them access to the solder and any parent folders. Depending on how far this gets nested, it could get messy (Parent -Sub -Sub - Sub)

I think the current user permissions works well IF the primary use of projectsend is for the admins to upload files for their clients. 

However, when the client has permission to upload, how can it be determined which folders they can upload to? Obviously, they can upload to any folder they create. BUT, presently they see ALL folders on the system, which includes folders that they do not have files in. That means they could upload, and continue to upload, to folders you might not intend for them to upload to. Should they only see folders they have files in, or should they see the parent folders as well?

WHAT IF the folders had permissions and could have more than one client be assigned to it?

This would allow a team or group from one organization to all see the files in a specific folder, much like the current assignment on files. However, administration would be easier because files simply need to be dropped in a folder to be shared. 

When a client wants to upload a file they would only see folders they created and any FOLDERS they have been assigned to. I think that would be a much faster data query for populating the drop down list as well. 

I am only writing these thoughts from a user interface perspective. I have no ability to write code like you! Thank you for what you have been doing!!

**Comment by Matani-Git** [comment]
Here’s the corrected version:

---
Hi @mbiolsi & @ignacionelson,

To address this issue, we need to review the `file_editor.php` file. A filter for public folders exists, but in my fix, I made all folders public to bypass the permission problem. This is because permissions are currently set for files, not folders.

I removed some lines in `file_editor.php` to prevent displaying all folders to clients. With this adjustment, a client can only select folders they created and will only be able to move files they own into those folders.

Additionally, I modified the `folders.php` file to include folders created by the logged-in client, even if those folders don't have any files assigned to them. This change ensures that clients can manage their files and folders more effectively.

To test this fix, replace the `folders.php` file located in `/includes/classes/` and the `file_editor.php` file located in `/includes/forms/`.


here are the files to replace and test ! 
[Folders.php.zip](https://github.com/user-attachments/files/17837536/Folders.php.zip)
[file_editor.php.zip](https://github.com/user-attachments/files/17837537/file_editor.php.zip)

Important!
My fix will not work with MySQL 5.7 because MySQL 5.7 doesn't support Common Table Expressions (CTEs) with the WITH RECURSIVE syntax, it will only work with MySQL 8 or MariaDB! Please refer to previous posts in issue #1333 regarding this folder issue 

S.Matani

**Comment by mbiolsi** [comment]
I did fine ONE issue. 

When I am logged in as a client, I DO see folders created by the admins. HOWEVER when I create a new folder it shows up at first, but it is not visible the next time I go to manage files. 

When I edited one of my files, I was able to see the folder I made as a client - and ONLY that folder. So that part worked as expected. HOWEVER, after adding the files to the folder the client cannot see the file because the client cannot see the folder. 

As an administrator I can see both the client folder and the client. file. 

REPLICATE:
- Log in as client
- create a folder
- click on any other menu item (view my files or upload)
- click back on Manage files - the folder is gone.

**Comment by Matani-Git** [comment]
Hi @mbiolsi 
i can not reproduce the issue! 

REPLICATE:

-    Log in as client
-   create a folder
-   click on any other menu item (view my files or upload)
-   click back on Manage files - the folder is gone. > for me i can see the created folder!

are you sure you did also replace the folders.php from my last post?

**Comment by Pauperx** [comment]
> The Projectsend release r1720 introduced the folder implementation. However, this feature is not very useful if clients cannot see the folders and their contents!
> 
> Until now, we have only used categories and category filters to reduce the file lists. With the new folder feature, the structure has become even better!
> 
> The folders in Projectsend are not traditional directories but virtual folders designed to improve organization for admins and clients. All files remain in the one directory, and the folder information is stored in the tbl_folders table.
> 
> **Problem**: When I upload new files as an admin or operator and assign these files to clients and client groups, the clients cannot see the files in the designated folders. They can only see files in the root folder.
> 
> Analysis: In the getFolders function within the Folders class, the publicorclient argument is not set. As a result, folders that contain files assigned to specific clients or client groups are not displayed.
> 
> **Solution:** Check the access level (level): if level = 0, it indicates a client; otherwise, it’s an admin or operator. If level ≠ 0, keep the behavior as it is. If level = 0, retrieve the client ID. Display all folders that contain files assigned to the client or to a group of which the client is a member. Set the folders as “public” so that clients can see the contents. The permissions remain attached to the files themselves, not to the folders. (I am not sure how the future folder functionality is planned, but this setup is perfect for our needs. I hope this will be available as an option in a future release.) Empty folders or folders without assigned files will be hidden from the client. If a client has a file in a subfolder, all parent directories up to the root will be rendered and shown.
> 
> **Fix**: The fix involves modifying the Folders class in projectsend/includes/Classes/Folders.php. Rename the old Folders.php file to Folders.php.bak, upload the new file, and test.
> 
> You can download the fix here: [Folders.zip](https://github.com/user-attachments/files/17704311/Folders.zip)

Thank you for this fix! Where can I find Folders.php to replace it? I do not see the path /projectsend/includes/Classes/ in my projectsend Docker folder on my NAS. I installed ProjectSend using Portainer. Hopefully some of you may be able to give a suggestion on this?

Thank you kindly.

**Comment by sinocaweb** [comment]
> Hi [@mbiolsi](https://github.com/mbiolsi) & [@ignacionelson](https://github.com/ignacionelson),
> 
> To address this issue, we need to review the `file_editor.php` file. A filter for public folders exists, but in my fix, I made all folders public to bypass the permission problem. This is because permissions are currently set for files, not folders.
> 
> I removed some lines in `file_editor.php` to prevent displaying all folders to clients. With this adjustment, a client can only select folders they created and will only be able to move files they own into those folders.
> 
> Additionally, I modified the `folders.php` file to include folders created by the logged-in client, even if those folders don't have any files assigned to them. This change ensures that clients can manage their files and folders more effectively.
> 
> To test this fix, replace the `folders.php` file located in `/includes/classes/` and the `file_editor.php` file located in `/includes/forms/`.
> 
> here are the files to replace and test ! [Folders.php.zip](https://github.com/user-attachments/files/17837536/Folders.php.zip) [file_editor.php.zip](https://github.com/user-attachments/files/17837537/file_editor.php.zip)
> 
> Important! My fix will not work with MySQL 5.7 because MySQL 5.7 doesn't support Common Table Expressions (CTEs) with the WITH RECURSIVE syntax, it will only work with MySQL 8 or MariaDB! Please refer to previous posts in issue [#1333](https://github.com/projectsend/projectsend/issues/1333) regarding this folder issue
> 
> S.Matani

it's Great!
there's just one problem... when client side clicks on "manage files" it returns error 500! Maybe I forgot something? does it happen to you too?

**Comment by WolfgangAndreas** [comment]
Hi, got the same issue in r1720 - Will this be fixed in a (near) future release? :-)  Thank you!
